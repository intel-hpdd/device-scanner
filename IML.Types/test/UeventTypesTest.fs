// Copyright (c) 2018 Intel Corporation. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

module IML.Types.UeventTypesTest

open UeventTypes

open Fable.Import.Jest
open IML.CommonLibrary
open Matchers
open Fixtures

[
  ("add", fixtures.add);
  ("add DM", fixtures.addDm);
  ("remove", fixtures.remove);
  ("MdRaid", fixtures.addMdRaid);
]
  |> List.map (fun ((name, x)) ->
    Test(name, fun () ->
      expect.assertions 2

      let decoded =
        x
        |> UEvent.udevDecoder
        |> Result.unwrap

      toMatchSnapshot decoded

      Map.ofList [
          (decoded.devpath, decoded);
        ]
        |> BlockDevices.encoder
        |> Thot.Json.Encode.encode 2
        |> toMatchSnapshot
    )
  )
  |> testList "decode / encode UEvents "
