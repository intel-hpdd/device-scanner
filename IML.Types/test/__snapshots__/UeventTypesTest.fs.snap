// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decode / encode UEvents  MdRaid 1`] = `
Object {
  "devname": "/dev/md0",
  "devpath": "/devices/virtual/block/md0",
  "devtype": "disk",
  "dmLvName": null,
  "dmMultipathDevpath": null,
  "dmSlaveMMs": Array [],
  "dmUUID": null,
  "dmVgName": null,
  "dmVgSize": null,
  "fsType": null,
  "fsUsage": null,
  "major": "9",
  "mdDevs": Array [
    "/dev/sda",
    "/dev/sdd",
  ],
  "mdUUID": "685b40ee:f2bc2028:f056f6d2:e292c910",
  "minor": "0",
  "model": null,
  "partEntryNumber": null,
  "paths": Array [
    "/dev/disk/by-id/md-name-lotus-32vm6:0",
    "/dev/disk/by-id/md-uuid-685b40ee:f2bc2028:f056f6d2:e292c910",
    "/dev/md0",
  ],
  "readOnly": false,
  "scsi80": null,
  "scsi83": null,
  "serial": null,
  "size": "41910272",
  "vendor": null,
}
`;

exports[`decode / encode UEvents  MdRaid 2`] = `
"{
  \\"/devices/virtual/block/md0\\": {
    \\"major\\": \\"9\\",
    \\"minor\\": \\"0\\",
    \\"paths\\": [
      \\"/dev/disk/by-id/md-name-lotus-32vm6:0\\",
      \\"/dev/disk/by-id/md-uuid-685b40ee:f2bc2028:f056f6d2:e292c910\\",
      \\"/dev/md0\\"
    ],
    \\"devName\\": \\"/dev/md0\\",
    \\"devPath\\": \\"/devices/virtual/block/md0\\",
    \\"devType\\": \\"disk\\",
    \\"idVendor\\": null,
    \\"idModel\\": null,
    \\"idSerial\\": null,
    \\"idFsType\\": null,
    \\"idFsUsage\\": null,
    \\"idPartEntryNumber\\": null,
    \\"size\\": \\"41910272\\",
    \\"scsi80\\": null,
    \\"scsi83\\": null,
    \\"isReadOnly\\": false,
    \\"dmSlaveMms\\": [],
    \\"dmVgSize\\": null,
    \\"mdDevices\\": [
      \\"/dev/sda\\",
      \\"/dev/sdd\\"
    ],
    \\"dmMultipathDevicePath\\": null,
    \\"dmLvName\\": null,
    \\"dmVgName\\": null,
    \\"dmUuid\\": null,
    \\"mdUuid\\": \\"685b40ee:f2bc2028:f056f6d2:e292c910\\"
  }
}"
`;

exports[`decode / encode UEvents  add 1`] = `
Object {
  "devname": "/dev/sdb1",
  "devpath": "/devices/pci0000:00/0000:00:01.1/ata1/host1/target1:0:0/1:0:0:0/block/sdb/sdb1",
  "devtype": "partition",
  "dmLvName": "swap",
  "dmMultipathDevpath": true,
  "dmSlaveMMs": Array [],
  "dmUUID": "LVM-pV8TgNKMJVNrolJgMhVwg4CAeFFAIMC83IU1hvimWWlvmd5xQddtMIqRtjwOuKTz",
  "dmVgName": "centos",
  "dmVgSize": null,
  "fsType": "LVM2_member",
  "fsUsage": "filesystem",
  "major": "8",
  "mdDevs": Array [],
  "mdUUID": null,
  "minor": "17",
  "model": "VBOX_HARDDISK",
  "partEntryNumber": 1,
  "paths": Array [
    "/dev/disk/by-id/ata-VBOX_HARDDISK_VB304a0a0f-15e93f07-part1",
    "/dev/disk/by-path/pci-0000:00:01.1-ata-1.0-part1",
    "/dev/sdb1",
  ],
  "readOnly": true,
  "scsi80": "80",
  "scsi83": "83",
  "serial": "VBOX_HARDDISK_VB304a0a0f-15e93f07",
  "size": "81784832",
  "vendor": null,
}
`;

exports[`decode / encode UEvents  add 2`] = `
"{
  \\"/devices/pci0000:00/0000:00:01.1/ata1/host1/target1:0:0/1:0:0:0/block/sdb/sdb1\\": {
    \\"major\\": \\"8\\",
    \\"minor\\": \\"17\\",
    \\"paths\\": [
      \\"/dev/disk/by-id/ata-VBOX_HARDDISK_VB304a0a0f-15e93f07-part1\\",
      \\"/dev/disk/by-path/pci-0000:00:01.1-ata-1.0-part1\\",
      \\"/dev/sdb1\\"
    ],
    \\"devName\\": \\"/dev/sdb1\\",
    \\"devPath\\": \\"/devices/pci0000:00/0000:00:01.1/ata1/host1/target1:0:0/1:0:0:0/block/sdb/sdb1\\",
    \\"devType\\": \\"partition\\",
    \\"idVendor\\": null,
    \\"idModel\\": \\"VBOX_HARDDISK\\",
    \\"idSerial\\": \\"VBOX_HARDDISK_VB304a0a0f-15e93f07\\",
    \\"idFsType\\": \\"LVM2_member\\",
    \\"idFsUsage\\": \\"filesystem\\",
    \\"idPartEntryNumber\\": 1,
    \\"size\\": \\"81784832\\",
    \\"scsi80\\": \\"80\\",
    \\"scsi83\\": \\"83\\",
    \\"isReadOnly\\": true,
    \\"dmSlaveMms\\": [],
    \\"dmVgSize\\": null,
    \\"mdDevices\\": [],
    \\"dmMultipathDevicePath\\": true,
    \\"dmLvName\\": \\"swap\\",
    \\"dmVgName\\": \\"centos\\",
    \\"dmUuid\\": \\"LVM-pV8TgNKMJVNrolJgMhVwg4CAeFFAIMC83IU1hvimWWlvmd5xQddtMIqRtjwOuKTz\\",
    \\"mdUuid\\": null
  }
}"
`;

exports[`decode / encode UEvents  add DM 1`] = `
Object {
  "devname": "/dev/sdb1",
  "devpath": "/devices/pci0000:00/0000:00:01.1/ata1/host1/target1:0:0/1:0:0:0/block/sdb/sdb1",
  "devtype": "disk",
  "dmLvName": "swap",
  "dmMultipathDevpath": true,
  "dmSlaveMMs": Array [
    "8:16",
    "8:32",
  ],
  "dmUUID": "LVM-KHoa9g8GBwQJMHjQtL77pGj6b9R1YWrlEDy4qFTQ3cgVnmyhy1zB2cJx2l5yE26D",
  "dmVgName": "centos",
  "dmVgSize": "21466447872B",
  "fsType": "LVM2_member",
  "fsUsage": "filesystem",
  "major": "8",
  "mdDevs": Array [],
  "mdUUID": null,
  "minor": "17",
  "model": "VBOX_HARDDISK",
  "partEntryNumber": 1,
  "paths": Array [
    "/dev/disk/by-id/ata-VBOX_HARDDISK_VB304a0a0f-15e93f07-part1",
    "/dev/disk/by-path/pci-0000:00:01.1-ata-1.0-part1",
    "/dev/sdb1",
  ],
  "readOnly": true,
  "scsi80": "80",
  "scsi83": "83",
  "serial": "VBOX_HARDDISK_VB304a0a0f-15e93f07",
  "size": "81784832",
  "vendor": null,
}
`;

exports[`decode / encode UEvents  add DM 2`] = `
"{
  \\"/devices/pci0000:00/0000:00:01.1/ata1/host1/target1:0:0/1:0:0:0/block/sdb/sdb1\\": {
    \\"major\\": \\"8\\",
    \\"minor\\": \\"17\\",
    \\"paths\\": [
      \\"/dev/disk/by-id/ata-VBOX_HARDDISK_VB304a0a0f-15e93f07-part1\\",
      \\"/dev/disk/by-path/pci-0000:00:01.1-ata-1.0-part1\\",
      \\"/dev/sdb1\\"
    ],
    \\"devName\\": \\"/dev/sdb1\\",
    \\"devPath\\": \\"/devices/pci0000:00/0000:00:01.1/ata1/host1/target1:0:0/1:0:0:0/block/sdb/sdb1\\",
    \\"devType\\": \\"disk\\",
    \\"idVendor\\": null,
    \\"idModel\\": \\"VBOX_HARDDISK\\",
    \\"idSerial\\": \\"VBOX_HARDDISK_VB304a0a0f-15e93f07\\",
    \\"idFsType\\": \\"LVM2_member\\",
    \\"idFsUsage\\": \\"filesystem\\",
    \\"idPartEntryNumber\\": 1,
    \\"size\\": \\"81784832\\",
    \\"scsi80\\": \\"80\\",
    \\"scsi83\\": \\"83\\",
    \\"isReadOnly\\": true,
    \\"dmSlaveMms\\": [
      \\"8:16\\",
      \\"8:32\\"
    ],
    \\"dmVgSize\\": \\"21466447872B\\",
    \\"mdDevices\\": [],
    \\"dmMultipathDevicePath\\": true,
    \\"dmLvName\\": \\"swap\\",
    \\"dmVgName\\": \\"centos\\",
    \\"dmUuid\\": \\"LVM-KHoa9g8GBwQJMHjQtL77pGj6b9R1YWrlEDy4qFTQ3cgVnmyhy1zB2cJx2l5yE26D\\",
    \\"mdUuid\\": null
  }
}"
`;

exports[`decode / encode UEvents  remove 1`] = `
Object {
  "devname": "/dev/sdb1",
  "devpath": "/devices/pci0000:00/0000:00:01.1/ata1/host1/target1:0:0/1:0:0:0/block/sdb/sdb1",
  "devtype": "partition",
  "dmLvName": null,
  "dmMultipathDevpath": null,
  "dmSlaveMMs": Array [],
  "dmUUID": null,
  "dmVgName": null,
  "dmVgSize": null,
  "fsType": null,
  "fsUsage": null,
  "major": "8",
  "mdDevs": Array [],
  "mdUUID": null,
  "minor": "17",
  "model": "VBOX_HARDDISK",
  "partEntryNumber": 1,
  "paths": Array [
    "/dev/disk/by-id/ata-VBOX_HARDDISK_VB304a0a0f-15e93f07-part1",
    "/dev/disk/by-path/pci-0000:00:01.1-ata-1.0-part1",
    "/dev/sdb1",
  ],
  "readOnly": null,
  "scsi80": null,
  "scsi83": null,
  "serial": "VBOX_HARDDISK_VB304a0a0f-15e93f07",
  "size": null,
  "vendor": null,
}
`;

exports[`decode / encode UEvents  remove 2`] = `
"{
  \\"/devices/pci0000:00/0000:00:01.1/ata1/host1/target1:0:0/1:0:0:0/block/sdb/sdb1\\": {
    \\"major\\": \\"8\\",
    \\"minor\\": \\"17\\",
    \\"paths\\": [
      \\"/dev/disk/by-id/ata-VBOX_HARDDISK_VB304a0a0f-15e93f07-part1\\",
      \\"/dev/disk/by-path/pci-0000:00:01.1-ata-1.0-part1\\",
      \\"/dev/sdb1\\"
    ],
    \\"devName\\": \\"/dev/sdb1\\",
    \\"devPath\\": \\"/devices/pci0000:00/0000:00:01.1/ata1/host1/target1:0:0/1:0:0:0/block/sdb/sdb1\\",
    \\"devType\\": \\"partition\\",
    \\"idVendor\\": null,
    \\"idModel\\": \\"VBOX_HARDDISK\\",
    \\"idSerial\\": \\"VBOX_HARDDISK_VB304a0a0f-15e93f07\\",
    \\"idFsType\\": null,
    \\"idFsUsage\\": null,
    \\"idPartEntryNumber\\": 1,
    \\"size\\": null,
    \\"scsi80\\": null,
    \\"scsi83\\": null,
    \\"isReadOnly\\": null,
    \\"dmSlaveMms\\": [],
    \\"dmVgSize\\": null,
    \\"mdDevices\\": [],
    \\"dmMultipathDevicePath\\": null,
    \\"dmLvName\\": null,
    \\"dmVgName\\": null,
    \\"dmUuid\\": null,
    \\"mdUuid\\": null
  }
}"
`;
